File: backend/app/models/pbti_result.py
# backend/app/models/pbti_result.py
from __future__ import annotations

from sqlalchemy import Integer, String, Boolean, ForeignKey, Index
from sqlalchemy.orm import Mapped, mapped_column, relationship
from sqlalchemy.dialects.mysql import JSON, BINARY

from .base import Base, TimestampMixin

class PBTIResult(Base, TimestampMixin):
    __tablename__ = "pbti_result"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)

    user_id: Mapped[bytes] = mapped_column(
        BINARY(16),
        ForeignKey("user.id", ondelete="CASCADE"),
        index=True,
        nullable=False,
    )

    # 0~100 점수
    temperature_score: Mapped[int] = mapped_column(Integer, nullable=False)
    texture_score: Mapped[int] = mapped_column(Integer, nullable=False)
    mood_score: Mapped[int] = mapped_column(Integer, nullable=False)
    nature_score: Mapped[int] = mapped_column(Integer, nullable=False)

    final_type: Mapped[str] = mapped_column(String(4), nullable=False)
    type_name: Mapped[str] = mapped_column(String(50), nullable=False)

    # [{"question_id": 1, "choice": 5, "axis": 1, "direction": 1, "score": 5}, ...]
    answers: Mapped[list[dict] | None] = mapped_column(JSON)

    is_active: Mapped[bool] = mapped_column(Boolean, default=False, nullable=False)

    user = relationship("User", backref="pbti_results")

    __table_args__ = (
        Index("ix_pbti_user_active", "user_id", "is_active"),
    )

File: lib/models/pbti_result.dart
// lib/models/pbti_result.dart
class PbtiAnswerModel {
  final int questionId;
  final int choice;
  final int score;

  PbtiAnswerModel({
    required this.questionId,
    required this.choice,
    required this.score,
  });

  factory PbtiAnswerModel.fromJson(Map<String, dynamic> json) {
    return PbtiAnswerModel(
      questionId: json['question_id'] as int? ?? 0,
      choice: json['choice'] as int? ?? 0,
      score: (json['score'] is double)
          ? (json['score'] as double).toInt()
          : (json['score'] as int? ?? 0),
    );
  }
}

class PbtiResultModel {
  final int temperatureScore;
  final int textureScore;
  final int moodScore;
  final int natureScore;
  final String finalType;
  final String typeName;
  final double confidence;

  /// 서버가 보내는 answers (optional)
  final List<PbtiAnswerModel> answers;

  PbtiResultModel({
    required this.temperatureScore,
    required this.textureScore,
    required this.moodScore,
    required this.natureScore,
    required this.finalType,
    required this.typeName,
    required this.confidence,
    required this.answers,
  });

  factory PbtiResultModel.fromJson(Map<String, dynamic> json) {
    int _safeInt(dynamic v) {
      if (v is int) return v;
      if (v is double) return v.toInt();
      return int.tryParse(v.toString()) ?? 0;
    }

    double _safeDouble(dynamic v) {
      if (v is double) return v;
      if (v is int) return v.toDouble();
      return double.tryParse(v.toString()) ?? 0.0;
    }

    List<PbtiAnswerModel> parsedAnswers = [];
    if (json['answers'] is List) {
      parsedAnswers = (json['answers'] as List)
          .map((e) => PbtiAnswerModel.fromJson(e))
          .toList();
    }

    return PbtiResultModel(
      temperatureScore: _safeInt(json['temperature_score']),
      textureScore: _safeInt(json['texture_score']),
      moodScore: _safeInt(json['mood_score']),
      natureScore: _safeInt(json['nature_score']),
      finalType: json['final_type'] as String? ?? '----',
      typeName: json['type_name'] as String? ?? '',
      confidence: _safeDouble(json['confidence']),
      answers: parsedAnswers,
    );
  }
}
